### ISSUE TITLE
Update example to use `Turbo` instead of `Turbolinks` (no longer under development)

### ISSUE THREAD

**Issue by lonroth** [Shopify/turbolinks-jwt-sample-app#25]
I would suggest updating this example repo to use `Turbo` (https://github.com/hotwired/turbo) instead of `Turbolinks` (https://github.com/turbolinks/turbolinks) since it's no longer under active development. It would also be great to see a complete tutorial using Hotwire (https://hotwire.dev/) since this is the future of Rails.

**Comment by kirillplatonov** [comment]
I created a sample repo with `Hotwire` & `Turbo` instead of `Turbolinks`: 
https://github.com/kirillplatonov/shopify-hotwire-sample

It's really fun working with `Hotwire` on Shopify Apps, especially with Streams.

**Comment by lonroth** [comment]
@kirillplatonov Awesome 👏 I've also created a Hotwire base app since I created the issue above and am currently working on a version using good old Sprockets instead of Webpack(er). :-) Happy to share when done if you're interested!

One difference from your implementation above and mine is how we pass the Bearer to the request. You use:
```
Turbo.setRequestInterceptor(async (request) => {
  const token = await getSessionToken(window.app);
  request.addHeader("Authorization", `Bearer ${token}`);
});
```

While I do this:
```
document.addEventListener("turbo:before-fetch-request", function (event) {
 var headers = event.detail.fetchOptions.headers;
 headers["Authorization"] = "Bearer " + window.sessionToken;
});
```

Both implementations seems to be working from what I can tell. Would you mind sharing why you settled on `setRequestInterceptor` instead of using `turbo:before-fetch-request` ? Hopefully it can also be useful for others reading our conversation 🙂

**Comment by dan-gamble** [comment]
@lonroth I think because it's not actual functionality of Turbo yet. You can see @kirillplatonov opened a PR for it here: https://github.com/hotwired/turbo/pull/177

**Comment by lonroth** [comment]
@dan-gamble Thanks for the PR link! I was just wondering why I hadn't seen `setRequestInterceptor ` before. I'll add a comment to the PR regarding this.

**Comment by kirillplatonov** [comment]
@lonroth yeah, I implemented interceptor for Turbo since I don't like the current solution with polling Shopify every 2s for token. Hopefully, this will be merged to Turbo. As for right now, I'm using it from a fork: https://github.com/kirillplatonov/shopify-hotwire-sample/blob/main/package.json#L5

The interceptor solution is inspired by Shopify's guide for Axios: https://shopify.dev/tutorials/authenticate-your-app-using-session-tokens#use-session-tokens-with-axios

Though, even with the interceptor, the solution is not perfect since we make an additional network request before any Turbo visit or form submission. It would be awesome having some kind of `expire_at` timestamp alongside with token in `getSessionToken` method. This way we could cache the token on the client-side and request token only when it's actually expired.

**Comment by lonroth** [comment]
@kirillplatonov Yeah, I do agree `setRequestInterceptor` is the better option here. Having `expire_at` timestamp alongside with token in `getSessionToken` method would also be great. Hopefully they will add something like this in the future.

**Comment by lonroth** [comment]
@kirillplatonov I added the comment on the PR before I got your comment here so if you want to comment there also let me know or I can remove the comment and we can keep the conversation here regarding Shopify if you prefer. Let me know. :)

**Comment by kirillplatonov** [comment]
@lonroth Let's keep this discussion here since it's more about Shopify JWT token's best practices and not about Turbo itself.

The reasons why I don't like the solution with polling:
1. It creates lots of unnecessary network requests to Shopify servers. If the user spent 10 min inside the app and clicked only a few links and submitted one form - we'll still make 300 redundant network requests for sessionToken. Not cool for both the client and server-side.
2. JavaScript is paused on inactive tabs in mobile Safari and in certain cases on desktop as well (when you switching the tab and the device is going asleep). So there's a chance that the token will be outdated by the moment when the user will restore the tab and click on links or submit a form. It's especially bad for forms since you can't retry the request when you reloaded the page with the fresh token. Though, I had this issue with old timer value (it was 50s). Now when it's 2s the chance that this will happen is low.

Overall I don't think it's critical and polling should also work fine with the current interval of 2s.

@NabeelAhsen @rezaansyed @brendonchen-shopify do you have any other considerations on that?

**Comment by lonroth** [comment]
@kirillplatonov Sounds good!👍 I'll remove the comment on the Turbo PR and let's focus there on getting it merged to Turbo `master` so that we can get the functionality needed. Then let's continue Shopify specific discussions here as suggested.

**Comment by rezaansyed** [comment]
@kirillplatonov I agree with you that the interceptor pattern is better due to the reasons you've listed. We can incorporate the example once that's supported once that's supported in Turbo.

**Comment by anduong** [comment]
@rezaansyed the Interceptor is merged to Turbo https://github.com/hotwired/turbo/pull/177. Can we now consider incorporate this idea into the example? cc @kirillplatonov

**Comment by kirillplatonov** [comment]
@anduong the interceptor in Turbo has been replaced with pausable event handler. You can use this repo as sample: https://github.com/kirillplatonov/shopify-hotwire-sample. It includes up to date Turbo JWT integration, App Bridge 2, Rails Request.JS integration.

**Comment by anduong** [comment]
btw @kirillplatonov I do see some of Monkey patches for ShopifyApp to set the expired time for the token in your example repo https://github.com/kirillplatonov/shopify-hotwire-sample. Though it works, it create extra things need to be aware of beside the standard behavior of the shopify app gem. If shopify app could also incorporate the expire time solution, it could be better.

**Comment by kirillplatonov** [comment]
@anduong Just removed this monkey patch. It's not needed anymore, since it's included in my `shopify_app` fork that this sample uses. Unfortunately, Shopify is not accepting PR's to `shopify_app` gem for a while so those changes can't be merged.

**Comment by kirillplatonov** [comment]
The fork also includes fixes for App Bridge 2 compatibility: https://github.com/kirillplatonov/shopify_app/commits/master

**Comment by anduong** [comment]
Thanks for sharing this @kirillplatonov I will have a look on your updated repo 👍
